File: mk/crc32c.py
#!/usr/bin/env python2
import sys


def _crc32c_table():
    # We need to use the bit-by-bit reversed polynomial.
    #  - CRC-32 (PKZIP) uses 0x04C11DB7, reversed to 0xEDB88320.
    #  - CRC-32C (Castagnoli) uses 0x1EDC6F41, reversed to 0x82F63B78.
    polynomial = 0x82F63B78
    table = [0] * 256
    for i in xrange(256):
        table[i] = i
        for j in xrange(8):
            lsb = table[i] & 1
            table[i] >>= 1
            if lsb:
                table[i] ^= polynomial
    return table


def crc32c(buffer):
    table = _crc32c_table()
    acc = 0xffffffff
    for b in buffer:
        acc = table[(acc & 0xff) ^ ord(b)] ^ (acc >> 8)
    acc ^= 0xffffffff
    return acc


def _main():
    if len(sys.argv) != 2 or sys.argv[1] in ("-h", "-help", "--help"):
        print("Usage: %s filename" % sys.argv[0])
        sys.exit(1)
    with open(sys.argv[1], "r") as f:
        data = f.read()
    print "%x" % crc32c(data)


if __name__ == "__main__":
    _main()

File: src/stage1/crc32c.rs
use core::prelude::*;

// A 1KiB-sized table used to compute the checksum.
struct Table {
    data: [u32; 256]
}

pub fn table() -> Table {
    // We need to use the bit-by-bit reversed polynomial.
    //  - CRC-32 (PKZIP) uses 0x04C11DB7, reversed to 0xEDB88320.
    //  - CRC-32C (Castagnoli) uses 0x1EDC6F41, reversed to 0x82F63B78.
    let polynomial = 0x82F63B78_u32;
    let mut table = Table { data: [0u32; 256] };
    for (i, cell) in table.data.iter_mut().enumerate() {
        let mut result = i as u32;
        for _ in range(0, 8) {
            let lsb = result & 1;
            result >>= 1;
            if lsb == 1 {
                result ^= polynomial;
            }
        }
        *cell = result;
    }
    table
}

pub fn compute(table: &Table, buffer: &[u8]) -> u32 {
    let mut acc = 0xffffffff_u32;
    for b in buffer.iter() {
        let lookup_value = unsafe {
            let index: u8 = *b ^ (acc as u8);
            *table.data.get_unchecked(index as usize)
        };
        acc = lookup_value ^ (acc >> 8);
    }
    acc ^ 0xffffffff
}
